Remove Postgres config from db admin machine

1pretz1 opened the pull request:
More info in the commits ->

Trello: https://trello.com/c/ewzHAr6A/56-remove-postgresql-primary-and-postgresql-standby-infrastructure-and-configuration

kevindew reviewed (commented):
Thanks for this Pete, I think there’s a few postgresql bits that’ll be left on the host

1pretz1 commented on modules/govuk/manifests/node/s_db_admin.pp:
```diff
@@ -76,13 +76,11 @@
 
   # include the common config/tooling required for our DB admin class
   class { '::govuk::nodes::postgresql_db_admin':
+    enabled             => false,
```
This was my thoughts originally but noticed other files using 'enabled'. Done!

1pretz1 commented on modules/govuk/manifests/nodes/postgresql_db_admin.pp:
```diff
@@ -53,27 +63,39 @@
-  }
+    # We don't actually want to run a local PostgreSQL server, just
+    # configure the RDS one
+    class { '::postgresql::server':
```
Ah ha, so it can. Have added the param `package_ensure`.

1pretz1 commented on modules/govuk/manifests/nodes/postgresql_db_admin.pp:
```diff
@@ -53,27 +63,39 @@
+    include ::govuk_postgresql::server::not_slave
+
+    # Ensure the client class is installed
+    class { '::govuk_postgresql::client': }
```
Ah okay, the ensure variable is now being passed to both of these packages which accept the param `package_ensure`.

1pretz1 commented:
Thank you for the review @kevindew, your comments have been very useful! Sorry, a lot of this has been guess work as we didn't know Puppet well enough. 

Think I've made the changes you've requested and some more which may be needed, could you take another look please? Thanks!

kevindew commented on modules/govuk/manifests/nodes/postgresql_db_admin.pp:
```diff
@@ -64,16 +69,21 @@
 
   service { 'postgresql':
-    ensure  => stopped,
+    ensure => stopped,
   }
 
-  include ::govuk_postgresql::server::not_slave
+  class { '::govuk_postgresql::server::not_slave':
+    ensure => $ensure,
+  }
```
I think we might have to put this bit within a conditional of if ensure is present. I'm expecting that `service 'postgresql' ...` may error if that service doesn't exist and it looks like the not_slave only runs queries on a postgres service so an `ensure => absent` on it is a bit misleading

1pretz1 replied:
Ah okay, good point. Have added a conditional here, instead of in the `not_slave` class.

kevindew commented on modules/govuk_postgresql/manifests/client.pp:
```diff
@@ -1,5 +1,23 @@
+  # installs libpq-dev package needed for pg gem
+  class { 'postgresql::lib::devel':
+    package_ensure => $ensure,
+  }
```
This looks great 👍

kevindew commented on modules/govuk_postgresql/manifests/server/not_slave.pp:
```diff
@@ -3,10 +3,20 @@
+    Postgresql::Server::Db             <| tag == $name |>
+    Postgresql::Server::Database_grant <| tag == $name |>
+    Govuk_postgresql::Extension        <| tag == $name |>
+  }
```
I don't think we should have an ensure here since it seems to just be queries this does (I think) rather than install software.

kevindew reviewed (commented):
Great stuff Pete. This looks almost there now to my eyes.

Have you had a chance to try it on integration? There's a part of me that wonders if all the assumptions I've made in guiding you hold and it'd be worth testing out. (I imagine we can undo any deployments by re-deploying the main branch)

1pretz1 commented:
Awesome - thanks for the informative review Kev!

I've not tried it out yet, as I wanted to wait until I was confident in the changes. However, I think we're there now and so will do tomorrow 🧪

kevindew commented on modules/govuk/manifests/nodes/postgresql_db_admin.pp:
```diff
@@ -64,16 +69,21 @@
 
-  include ::govuk_postgresql::server::not_slave
+  if $ensure == present {
+    include ::govuk_postgresql::server::not_slave
```
I think you'll want the `service { 'postgresql'` block in this conditional too. I don't know how that will behave if the service is removed.

1pretz1 replied:
Good shout!

kevindew commented:
> Awesome - thanks for the informative review Kev!
> 
> I've not tried it out yet, as I wanted to wait until I was confident in the changes. However, I think we're there now and so will do tomorrow 🧪

Great, aside from my last comment this looks great. Good luck with the integration run, hopefully I've not given you any bad advice.

1pretz1 commented:
I tried this out on integration and it mostly worked.

For the app db admin machines, everything works fine and all the Postgres resources and config remain. For the generic db admin machine most of the resources were removed, but not all. The client software was removed and most of the dirs / files were deleted (`/root/.pgpass`, `/etc/postgresql/`, `/var/lib/postgresql/`). However the `/etc/init.d/postgresql` still persisted and thus upstart still recognised postgresql as a service.

Digging into the puppet output, everything seems to be going fine until `Error: Cannot create /var/lib/postgresql/9.6/main; parent directory /var/lib/postgresql/9.6 does not exist`. It seems that something is still trying to add something to the Postgres 9.6 directory. The next line is: `Error: /Stage[main]/Postgresql::Server::Initdb/File[/var/lib/postgresql/9.6/main]/ensure: change from absent to directory failed: Cannot create /var/lib/postgresql/9.6/main; parent directory /var/lib/postgresql/9.6 does not exist` which explains why the init script remained.

```
Notice: /Stage[main]/Postgresql::Client/Package[postgresql-client]/ensure: removed

Notice: /Stage[main]/Postgresql::Lib::Devel/Package[postgresql-devel]/ensure: removed

Info: Computing checksum on file /root/.pgpass

Info: /Stage[main]/Govuk::Nodes::Postgresql_db_admin/File[/root/.pgpass]: Filebucketed /root/.pgpass to puppet with sum 45024caa072ee16f75ff2ed78231f7f8

Notice: /Stage[main]/Govuk::Nodes::Postgresql_db_admin/File[/root/.pgpass]/ensure: removed

Info: Computing checksum on file /etc/apt/sources.list.d/postgresql.list

Info: /Stage[main]/Govuk::Nodes::Postgresql_db_admin/Apt::Source[postgresql]/Apt::Setting[list-postgresql]/File[/etc/apt/sources.list.d/postgresql.list]: Filebucketed /etc/apt/sources.list.d/postgresql.list to puppet with sum 46696071f38bc9b07946afc1626f6ffc

Notice: /Stage[main]/Govuk::Nodes::Postgresql_db_admin/Apt::Source[postgresql]/Apt::Setting[list-postgresql]/File[/etc/apt/sources.list.d/postgresql.list]/ensure: removed

Info: /Stage[main]/Govuk::Nodes::Postgresql_db_admin/Apt::Source[postgresql]/Apt::Setting[list-postgresql]/File[/etc/apt/sources.list.d/postgresql.list]: Scheduling refresh of Class[Apt::Update]
Info: Class[Apt::Update]: Scheduling refresh of Exec[apt_update]

Notice: /Stage[main]/Apt::Update/Exec[apt_update]: Triggered 'refresh' from 1 events

Notice: /Stage[main]/Postgresql::Server::Install/Package[postgresql-server]/ensure: ensure changed '9.6.9-2.pgdg14.04+1' to 'purged' [server removed]

Error: Cannot create /var/lib/postgresql/9.6/main; parent directory /var/lib/postgresql/9.6 does not exist

Error: /Stage[main]/Postgresql::Server::Initdb/File[/var/lib/postgresql/9.6/main]/ensure: change from absent to directory failed: Cannot create /var/lib/postgresql/9.6/main; parent directory /var/lib/postgresql/9.6 does not exist

Notice: /Stage[main]/Postgresql::Server::Config/Postgresql::Server::Config_entry[data_directory]/Exec[postgresql_stop_data_directory]: Dependency File[/var/lib/postgresql/9.6/main] has failures: true

Warning: /Stage[main]/Postgresql::Server::Config/Postgresql::Server::Config_entry[data_directory]/Exec[postgresql_stop_data_directory]: Skipping because of failed dependencies

Notice: /Stage[main]/Postgresql::Server::Config/Postgresql::Server::Pg_hba_rule[deny access to postgresql user]/Concat::Fragment[pg_hba_rule_deny access to postgresql user]/Concat_fragment[pg_hba_rule_deny access to postgresql user]: Dependency File[/var/lib/postgresql/9.6/main] has failures: true
```

kevindew commented:
Thanks Pete, looks like you're super close 🎉 

I'm wondering if the problem is this line: https://github.com/alphagov/govuk-puppet/blob/aa6e626ff8ab2c5e50d8b72e2c1dce2cc32497f8/modules/govuk/manifests/node/s_db_admin.pp#L88

(which probably should have been deleted in an earlier clean up but must have been missed)

That class has https://github.com/alphagov/govuk-puppet/blob/064405329eb2cb78290e5ef000a27bd9c821770b/modules/govuk/manifests/apps/ckan/db.pp#L31 which I imagine has a bunch of postgresql requires hidden away.

kevindew commented:
Superseded by https://github.com/alphagov/govuk-puppet/pull/11506
